File: apps/mobile/lib/utils/recent_pr.dart
import '../i18n/app_strings.dart';
import '../models/api_models.dart';

/// A new personal record detected by walking sessions in time order.
class RecentPr {
  final String exerciseName;
  final double kg;
  final int reps;
  final double e1rm;
  final DateTime when;

  const RecentPr({
    required this.exerciseName,
    required this.kg,
    required this.reps,
    required this.e1rm,
    required this.when,
  });
}

double _epley(double kg, int reps) => kg * (1 + reps / 30.0);

/// Scan sessions chronologically and return the most recent PR within
/// [withinDays]. A PR for an exercise = the session's max e1RM exceeds the
/// previous all-time max for that exercise by more than 0.5kg (filters out
/// floating-point churn from identical kg/reps).
RecentPr? findRecentPr(List<WorkoutSession> sessions,
    {int withinDays = 7}) {
  // Sort oldest → newest so the running max is built up correctly.
  final sorted = [...sessions]..sort((a, b) {
      final da = a.finishedAt ?? a.startedAt;
      final db = b.finishedAt ?? b.startedAt;
      if (da == null && db == null) return 0;
      if (da == null) return -1;
      if (db == null) return 1;
      return da.compareTo(db);
    });

  final allTimeMax = <String, double>{}; // exerciseId → best e1RM ever
  RecentPr? latest;
  final now = DateTime.now();

  for (final session in sorted) {
    final when = session.finishedAt ?? session.startedAt;
    if (when == null) continue;
    for (final ex in session.exercises) {
      if (ex.exerciseId.isEmpty || ex.sets.isEmpty) continue;
      // Best top-set in this session.
      double bestE1rm = 0;
      double bestKg = 0;
      int bestReps = 0;
      for (final set in ex.sets) {
        if (!set.done || set.reps <= 0) continue;
        final e1rm = _epley(set.kg, set.reps);
        if (e1rm > bestE1rm) {
          bestE1rm = e1rm;
          bestKg = set.kg;
          bestReps = set.reps;
        }
      }
      if (bestE1rm <= 0) continue;
      final prior = allTimeMax[ex.exerciseId] ?? 0;
      if (bestE1rm > prior + 0.5) {
        // It's a PR. Only surface it if it falls inside the window.
        if (now.difference(when).inDays <= withinDays) {
          latest = RecentPr(
            exerciseName: ex.name,
            kg: bestKg,
            reps: bestReps,
            e1rm: bestE1rm,
            when: when,
          );
        }
        allTimeMax[ex.exerciseId] = bestE1rm;
      }
    }
  }
  return latest;
}

/// Returns the set of session ids that contained at least one PR. Walks
/// sessions chronologically, keeps a running max e1RM per exercise, and flags
/// any session whose top set exceeds the previous max by >0.5 kg.
Set<String> sessionsWithPr(List<WorkoutSession> sessions) {
  final sorted = [...sessions]..sort((a, b) {
      final da = a.finishedAt ?? a.startedAt;
      final db = b.finishedAt ?? b.startedAt;
      if (da == null && db == null) return 0;
      if (da == null) return -1;
      if (db == null) return 1;
      return da.compareTo(db);
    });
  final allTimeMax = <String, double>{};
  final flagged = <String>{};
  for (final session in sorted) {
    bool prThisSession = false;
    for (final ex in session.exercises) {
      if (ex.exerciseId.isEmpty || ex.sets.isEmpty) continue;
      double bestE1rm = 0;
      for (final set in ex.sets) {
        if (!set.done || set.reps <= 0) continue;
        final e1rm = _epley(set.kg, set.reps);
        if (e1rm > bestE1rm) bestE1rm = e1rm;
      }
      if (bestE1rm <= 0) continue;
      final prior = allTimeMax[ex.exerciseId] ?? 0;
      if (bestE1rm > prior + 0.5) {
        prThisSession = true;
        allTimeMax[ex.exerciseId] = bestE1rm;
      }
    }
    if (prThisSession && session.id.isNotEmpty) flagged.add(session.id);
  }
  return flagged;
}

/// "Today" / "Yesterday" / "N days ago" — short relative date label for
/// the PR banner. Looks up the canonical strings from the i18n dict so it
/// flips with the language toggle.
String relativeDayLabel(DateTime when, {DateTime? now}) {
  final n = now ?? DateTime.now();
  final today = DateTime(n.year, n.month, n.day);
  final that = DateTime(when.year, when.month, when.day);
  final diff = today.difference(that).inDays;
  if (diff <= 0) return t('date.today');
  if (diff == 1) return t('date.yesterday');
  return tFmt('date.days_ago', {'n': diff});
}

File: apps/native-mobile/src/utils/recentPr.ts
// Ported 1:1 from apps/mobile/lib/utils/recent_pr.dart.
import { t } from "../i18n";
import { WorkoutSession } from "../models/apiModels";

export class RecentPr {
  constructor(
    public exerciseName: string,
    public kg: number,
    public reps: number,
    public e1rm: number,
    public when: Date,
  ) {}
}

function epley(kg: number, reps: number): number {
  return kg * (1 + reps / 30.0);
}

function sortChrono(sessions: WorkoutSession[]): WorkoutSession[] {
  return [...sessions].sort((a, b) => {
    const da = a.finishedAt ?? a.startedAt;
    const db = b.finishedAt ?? b.startedAt;
    if (da == null && db == null) return 0;
    if (da == null) return -1;
    if (db == null) return 1;
    return da.getTime() - db.getTime();
  });
}

/** Most recent PR within [withinDays] (e1RM exceeds prior max by >0.5kg). */
export function findRecentPr(sessions: WorkoutSession[], withinDays = 7): RecentPr | null {
  const sorted = sortChrono(sessions);
  const allTimeMax: Record<string, number> = {};
  let latest: RecentPr | null = null;
  const now = new Date();

  for (const session of sorted) {
    const when = session.finishedAt ?? session.startedAt;
    if (when == null) continue;
    for (const ex of session.exercises) {
      if (ex.exerciseId.length === 0 || ex.sets.length === 0) continue;
      let bestE1rm = 0;
      let bestKg = 0;
      let bestReps = 0;
      for (const s of ex.sets) {
        if (!s.done || s.reps <= 0) continue;
        const e1rm = epley(s.kg, s.reps);
        if (e1rm > bestE1rm) {
          bestE1rm = e1rm;
          bestKg = s.kg;
          bestReps = s.reps;
        }
      }
      if (bestE1rm <= 0) continue;
      const prior = allTimeMax[ex.exerciseId] ?? 0;
      if (bestE1rm > prior + 0.5) {
        const diffDays = Math.floor((now.getTime() - when.getTime()) / 86400000);
        if (diffDays <= withinDays) {
          latest = new RecentPr(ex.name, bestKg, bestReps, bestE1rm, when);
        }
        allTimeMax[ex.exerciseId] = bestE1rm;
      }
    }
  }
  return latest;
}

/** Set of session ids that contained at least one PR. */
export function sessionsWithPr(sessions: WorkoutSession[]): Set<string> {
  const sorted = sortChrono(sessions);
  const allTimeMax: Record<string, number> = {};
  const flagged = new Set<string>();
  for (const session of sorted) {
    let prThisSession = false;
    for (const ex of session.exercises) {
      if (ex.exerciseId.length === 0 || ex.sets.length === 0) continue;
      let bestE1rm = 0;
      for (const s of ex.sets) {
        if (!s.done || s.reps <= 0) continue;
        const e1rm = epley(s.kg, s.reps);
        if (e1rm > bestE1rm) bestE1rm = e1rm;
      }
      if (bestE1rm <= 0) continue;
      const prior = allTimeMax[ex.exerciseId] ?? 0;
      if (bestE1rm > prior + 0.5) {
        prThisSession = true;
        allTimeMax[ex.exerciseId] = bestE1rm;
      }
    }
    if (prThisSession && session.id.length > 0) flagged.add(session.id);
  }
  return flagged;
}

/** "Today" / "Yesterday" / "N days ago" — short relative date label. */
export function relativeDayLabel(when: Date, now?: Date): string {
  const n = now ?? new Date();
  const today = new Date(n.getFullYear(), n.getMonth(), n.getDate());
  const that = new Date(when.getFullYear(), when.getMonth(), when.getDate());
  const diff = Math.floor((today.getTime() - that.getTime()) / 86400000);
  if (diff <= 0) return t("date.today");
  if (diff === 1) return t("date.yesterday");
  return t("date.days_ago", { n: diff });
}
